README
This wrapper is based on gtkglarea 1.99.0.  You can get the source from 

ftp://ftp.gnome.org/pub/GNOME/sources/gtkglarea/1.99/gtkglarea-1.99.0.tar.bz2

Mailing list at:

http://lists.colliertech.org/cgi-bin/mailman/listinfo/gtkglarea-dev


ChangeLog
2009-07-17T21:59  Sam Hocevar <sam@zoy.org>

	* gtkglarea/gtkglarea-sharp.metadata: mark GLArea.Begingl() and
        GLArea.Endgl() as hidden so that they do not appear in the API. These
	functions have been deprecated for 10 years now.

2009-07-17T21:37  Sam Hocevar <sam@zoy.org>

	* gtkglarea/gtkglarea-sharp.metadata: mark GLArea.Size() as new in
        order to override GtkDrawingArea.Size().

2009-07-17T02:14  Sam Hocevar <sam@zoy.org>

	Overview: more complete and consistent Gapi detection.

	* configure.ac:
	- Check for gapi-parser / gapi2-parser.
	- Rename FIXUP and CODEGEN to GAPI_FIXUP and GAPI_CODEGEN
	* sources/Makefile.am gtkglarea/Makefile.am:
	- Reflect changes made to configure.ac.
	- Call $(GAPI_FIXUP) and $(GAPI_PARSER) instead of gapi2-fixup and
	gapi2-parser.

2009-07-17T01:53  Sam Hocevar <sam@zoy.org>

	Remove everything gtkglext until we really decide to support it; as
	of now, there are only empty files.

2009-07-17T01:48  Sam Hocevar <sam@zoy.org>

	Regenerate gtkglarea/generated/*.cs using a newer gapi.

2008-04-15T1507  Mike Kestner <mkestner@novell.com>

	* autogen.sh:  Patch from Sam Hocevar submitted to gtk#
	bug #371845.  Supports automake 1.10+.

2007-12-20T1709  C.J. Adams-Collier  <cjac@colliertech.org>

	Overview (from Sam Hocevar): patch for gtkglarea-sharp adds
	gtk_gl_area_share_new() to the list of bindings to allow sharing
	the GL context. This way one can share textures and display lists
	between GL contexts.

	* gtkglarea/GLArea.custom
	- created c'tor which accepts a share list
	* gtkglarea/generated/GLArea.cs
	- code generated from above

2007-12-20T1709  C.J. Adams-Collier  <cjac@colliertech.org>

	Overview (from Sam Hocevar): The gtkglarea-sharp package fails to
	properly detect the gapi tools, causing "make dist" to always
	fail.

	* configure.ac
	- changing positive check for xyes to negative check for xno

2007-12-20T1705  C.J. Adams-Collier  <cjac@colliertech.org>
	
	Overview (from Sam Hocevar): Minor issue in gtkglarea-sharp's
	PickWidget.cs: the right mouse button is ignored upon press, but
	not upon release, leading to weird behaviour.

	* examples/pick-object/PickWidget.cs
	- ignoring right mouse button on release

2007-12-20T1649  C.J. Adams-Collier  <cjac@colliertech.org>

	Overview (from Sam Hocevar): The gtkglarea-sharp build process
	expects a Tao.OpenGl.pc file, but such a file doesn't exist
	anywhere AFAIK. The Tao Debian packages use tao-opengl.pc for
	consistency so I suggest using the same naming scheme
	(see also http://www.taoframework.com/node/432 for a discussion
	with the Tao maintainers about this).

	* configure.ac, examples/Include.am, examples/lwo/Makefile.am
	- s/Tao.OpenGl/tao-opengl/g
	
2007-12-20T1642  C.J. Adams-Collier  <cjac@colliertech.org>

	Overview: Applied some patches supplied by Sam Hocevar
	
	* AUTHORS
	- added Sam Hocevar
	* configure.ac
	- put a $ in front of a variable :)
	* examples/shapes/Teapot.cs, examples/rotate-buttons/OldTeapot.cs
	- changed data structures to conform to new API

2007-04-23T06:35  C.J. Adams-Collier  <cjac@colliertech.org>
	* gtkglarea/Makefile.am
	- removed generated-stamp and generated/* from CLEANFILES
	- added generated/* and generated-stamp to EXTRA_DIST so that the
	  build no longer requires that the builder have gapi
	* configure.ac
	- did a bunch of clean-up
	* gtkglarea/generated
	- Added gtkglarea/generated/*.cs to svn so they don't need to be
	  generated
	* TODO
	- added some things I'd like to do/see done

2007-04-09T20:54  C.J. Adams-Collier  <cjac@colliertech.org>
	- cleaned up example makefiles using automake's include
	* examples/.../Makefile.am
	- re-factored to take advantage of new Include.am files
	* examples/Include.am
	- placed common MCS_ARGS, MONO_PATH, CLEANFILES, EXTRA_DIST,
	  $(ASSEMBLY): target, run: target, and all: target here
	* examples/glade/Include.am,
	* examples/shapes/Include.am,
	* examples/util/Include.am
	- placed MCS_ARGS, MONO_PATH, and RESOURCE_FILES specific to these
	  libs here
	* README
	- updated
	* configure.ac
	- added AC_SUBST for variables that need to be set for the
	  Include.am files to function correctly
	- added a HAS_TAO_OPENGL conditional to help decide whether to
	  build examples
	* Makefile.am
	- using HAS_TAO_OPENGL to decide whether to build examples/*
	* gtkglarea/Makefile.am
	- replaced duplicated strings with variables
	- cleaned up a bit
	
	
2007-04-08T20:32  C.J. Adams-Collier  <cjac@colliertech.org>
	* examples/pick-object/Engine.cs
	- made comments more specific
	- cube now throbs, too

2007-04-08T08:56  C.J. Adams-Collier  <cjac@colliertech.org>
	* */Makefile.am
	- ensured that 'make distcheck' completes successfully
	- building .dll files for classes used outside of each particular
	  directory and using -lib: and -r: instead of mentioning the .cs
	  explicitly (Teapot.cs, GladeExample.cs, Trackball.cs)
	- using $(addprefix $(srcdir)/, $(SOURCE_FILES)) instead of
	  $(SOURCE_FILES) so that builddir can differ from srcdir
	- added *.mdb to CLEANFILES
	- using -lib:@top_builddir@ instead of -lib:$(top_srcdir) so that
	  source can be built from somewhere other than the source dir
	* configure.ac
	- using more modern version of AC_INIT
	- explicitly using tar-ustar and requiring automake 1.9
	- bumped version to 0.0.16
	- using AC_CONFIG_FILES to specify files
	- using non-deprecated AC_OUTPUT macro

2007-04-07T13:39  C.J. Adams-Collier  <cjac@colliertech.org>
	- modified svn:ignore properties
	* examples/util/GLObjectBase.cs
	- set defaults for translation matrix values
	- set defaults for scale matrix values
	- added some docs describing what trans and scale matrix values
	  are being set to
	- setting scale matrix correctly
	* examples/util/GLWidget.cs
	- replaced iterator with foreach
	* examples/pick-object/ObjectPickExample.mdp
	- removed dependency on trackball assembly
	- replaced Tao.OpenGl 0.0.0.0 and Tao.OpenGl.Glu with Tao.OpenGl
	  2.1.1.0
	* examples/pick-object/Engine.cs

examples/util/Teapot.cs
// created on 11/25/2005 at 11:13 AM
using Tao.OpenGl;
using gl=Tao.OpenGl.Gl;
using System;
using System.Collections;

namespace GtkGL {
	public class Teapot : GtkGL.GLObjectBase,GtkGL.IGLObject { 
		  /* 
		   * Teapot
		   */

		  /* Rim, body, lid, and bottom data must be reflected in x and
		     y; handle and spout data across the y axis only.  */

		  static int[,] patchdata =
		    {
		      /* rim */
		      {102, 103, 104, 105, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15},
		      /* body */
		      {12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27},
		      {24, 25, 26, 27, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40},
		      /* lid */
		      {96, 96, 96, 96, 97, 98, 99, 100, 101, 101, 101, 101, 0, 1, 2, 3,},
		      {0, 1, 2, 3, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117},
		      /* bottom */
		      {118, 118, 118, 118, 124, 122, 119, 121, 123, 126, 125, 120, 40, 39, 38, 37},
		      /* handle */
		      {41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56},
		      {53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 28, 65, 66, 67},
		      /* spout */
		      {68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83},
		      {80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95}
		    };
		  /* *INDENT-OFF* */

		  static double[,] cpdata =
		    {
		      {0.2, 0, 2.7}, {0.2, -0.112, 2.7}, {0.112, -0.2, 2.7},
		      {0,-0.2, 2.7}, {1.3375, 0, 2.53125}, {1.3375, -0.749, 2.53125},
		      {0.749, -1.3375, 2.53125}, {0, -1.3375, 2.53125}, {1.4375,0, 2.53125},
		      {1.4375, -0.805, 2.53125}, {0.805, -1.4375,2.53125}, {0, -1.4375, 2.53125},
		      {1.5, 0, 2.4}, {1.5, -0.84, 2.4}, {0.84, -1.5, 2.4},
		      {0, -1.5, 2.4}, {1.75, 0, 1.875},
		      {1.75, -0.98, 1.875}, {0.98, -1.75, 1.875}, {0, -1.75, 1.875},
		      {2, 0, 1.35}, {2, -1.12, 1.35}, {1.12, -2, 1.35},
		      {0, -2, 1.35}, {2, 0, 0.9}, {2, -1.12, 0.9},
		      {1.12, -2, 0.9}, {0, -2, 0.9}, {-2, 0, 0.9},
		      {2, 0, 0.45}, {2, -1.12, 0.45}, {1.12, -2, 0.45},
		      {0, -2, 0.45}, {1.5, 0, 0.225},
		      {1.5, -0.84, 0.225}, {0.84, -1.5, 0.225}, {0, -1.5, 0.225},
		      {1.5, 0, 0.15}, {1.5, -0.84, 0.15}, {0.84, -1.5, 0.15},
		      {0, -1.5, 0.15}, {-1.6, 0, 2.025}, {-1.6, -0.3, 2.025},
		      {-1.5, -0.3, 2.25}, {-1.5, 0, 2.25}, {-2.3, 0, 2.025},
		      {-2.3, -0.3, 2.025}, {-2.5, -0.3, 2.25}, {-2.5, 0, 2.25},
		      {-2.7, 0, 2.025}, {-2.7, -0.3, 2.025}, {-3, -0.3, 2.25},
		      {-3, 0, 2.25}, {-2.7, 0, 1.8}, {-2.7, -0.3, 1.8},
		      {-3, -0.3, 1.8}, {-3, 0, 1.8}, {-2.7, 0, 1.575},
		      {-2.7, -0.3, 1.575}, {-3, -0.3, 1.35}, {-3, 0, 1.35},
		      {-2.5, 0, 1.125}, {-2.5, -0.3, 1.125}, {-2.65, -0.3, 0.9375},
		      {-2.65, 0, 0.9375}, {-2, -0.3, 0.9}, {-1.9, -0.3, 0.6},
		      {-1.9, 0, 0.6}, {1.7, 0, 1.425}, {1.7, -0.66, 1.425},
		      {1.7, -0.66, 0.6}, {1.7, 0, 0.6}, {2.6, 0, 1.425},
		      {2.6, -0.66, 1.425}, {3.1, -0.66, 0.825}, {3.1, 0, 0.825},
		      {2.3, 0, 2.1}, {2.3, -0.25, 2.1},
		      {2.4, -0.25, 2.025}, {2.4, 0, 2.025}, {2.7, 0, 2.4},
		      {2.7, -0.25, 2.4}, {3.3, -0.25, 2.4}, {3.3, 0, 2.4},
		      {2.8, 0, 2.475}, {2.8, -0.25, 2.475}, {3.525, -0.25, 2.49375},
		      {3.525, 0, 2.49375}, {2.9, 0, 2.475}, {2.9, -0.15, 2.475},
		      {3.45, -0.15, 2.5125}, {3.45, 0, 2.5125}, {2.8, 0, 2.4},
		      {2.8, -0.15, 2.4}, {3.2, -0.15, 2.4}, {3.2, 0, 2.4},
		      {0, 0, 3.15}, {0.8, 0, 3.15}, {0.8, -0.45, 3.15},
		      {0.45, -0.8, 3.15}, {0, -0.8, 3.15}, {0, 0, 2.85},
		      {1.4, 0, 2.4}, {1.4, -0.784, 2.4}, {0.784, -1.4, 2.4},
		      {0, -1.4, 2.4}, {0.4, 0, 2.55}, {0.4, -0.224, 2.55},
		      {0.224, -0.4, 2.55}, {0, -0.4, 2.55}, {1.3, 0, 2.55}, {1.3, -0.728, 2.55}, {0.728, -1.3, 2.55}, {0, -1.3, 2.55}, {1.3, 0, 2.4}, {1.3, -0.728, 2.4},
		      {0.728, -1.3, 2.4}, {0, -1.3, 2.4}, {0, 0, 0}, {1.425, -0.798, 0}, {1.5, 0, 0.075}, {1.425, 0, 0}, {0.798, -1.425, 0}, {0, -1.5, 0.075}, {0, -1.425, 0}, {1.5, -0.84, 0.075},
		      {0.84, -1.5, 0.075}
		    };

		  static float[,,] tex =
		    {
		      { {0, 0},
			{1, 0}},
		      { {0, 1},
			{1, 1}}
		    };

		  /* *INDENT-ON* */
		  
		  	static float[][]
		  		cachedP = new float[10][],
			  	cachedQ = new float[10][],
			  	cachedR = new float[10][],
			  	cachedS = new float[10][];

		  	private void DoTeaPot(int grid, double scale, int type)
		  	{
			
			    gl.glPushAttrib(gl.GL_ENABLE_BIT | gl.GL_EVAL_BIT);
			    gl.glEnable(gl.GL_AUTO_NORMAL);
			    gl.glEnable(gl.GL_NORMALIZE);
			    gl.glEnable(gl.GL_MAP2_VERTEX_3);
			    gl.glEnable(gl.GL_MAP2_TEXTURE_COORD_2);
			    gl.glPushMatrix();
			    gl.glRotatef(270.0f, 1.0f, 0.0f, 0.0f);
			    gl.glScalef((float)(0.5 * scale), (float)(0.5 * scale), (float)(0.5 * scale));
		    	gl.glTranslatef(0.0f, 0.0f, -1.5f);
		    
		    	float[,,]
		    		p = new float[4,4,3],
		    		q = new float[4,4,3],
		    		r = new float[4,4,3],
					s = new float[4,4,3];
		    	int i, j, k, l;
			    
			   	for (i = 0; i < 10; i++) {
			   	
				   	if(cachedP[i] == null){
				      	for (j = 0; j < 4; j++) {
							for (k = 0; k < 4; k++) {
					  			for (l = 0; l < 3; l++) {
					    			p[j,k,l] = (float)cpdata[patchdata[i,j * 4 + k],l];
					    			q[j,k,l] = (float)cpdata[patchdata[i,j * 4 + (3 - k)],l];
					    			if (l == 1)
						      			q[j,k,l] *= -1.0f;
					    			if (i < 6) {
					      				r[j,k,l] = (float)cpdata[patchdata[i,j * 4 + (3 - k)],l];
					      				if (l == 0)
											r[j,k,l] *= -1.0f;
					      				s[j,k,l] = (float)cpdata[patchdata[i,j * 4 + k],l];
					      				if (l == 0)
											s[j,k,l] *= -1.0f;
					      				if (l == 1)
											s[j,k,l] *= -1.0f;
						    		}
					  			}
							}
		    	  		}
			          
			    	  	cachedP[i] = new float[4*4*3];
		    		  	cachedQ[i] = new float[4*4*3];
		    		  	cachedR[i] = new float[4*4*3];
		    		  	cachedS[i] = new float[4*4*3];

				      	int pos;
			    	  	pos = 0;
						for(int I = 0; I < 4; I++)
			    	  		for(int J = 0; J < 4; J++)
			    	  			for(int K = 0; K < 3; K++,pos++){
					  				cachedP[i][pos] = p[I,J,K];
			    	  				cachedQ[i][pos] = q[I,J,K];
			    	  				cachedR[i][pos] = r[I,J,K];
			    	  				cachedS[i][pos] = s[I,J,K];
			    	  			}    	  			
		    	  	}
		    	  			
		    		gl.glMap2f(gl.GL_MAP2_TEXTURE_COORD_2, 0, 1, 2, 2, 0, 1, 4, 2, ref tex[0,0,0]);
		    		gl.glMap2f(gl.GL_MAP2_VERTEX_3, 0, 1, 3, 4, 0, 1, 12, 4, cachedP[i]);
		    		gl.glMapGrid2f(grid, 0.0f, 1.0f, grid, 0.0f, 1.0f);
		    		gl.glEvalMesh2(type, 0, grid, 0, grid);
		    		gl.glMap2f(gl.GL_MAP2_VERTEX_3, 0, 1, 3, 4, 0, 1, 12, 4, cachedQ[i]);
		    		gl.glEvalMesh2(type, 0, grid, 0, grid);
		    		if (i < 6) {
						gl.glMap2f(gl.GL_MAP2_VERTEX_3, 0, 1, 3, 4, 0, 1, 12, 4, cachedR[i]);
						gl.glEvalMesh2(type, 0, grid, 0, grid);
						gl.glMap2f(gl.GL_MAP2_VERTEX_3, 0, 1, 3, 4, 0, 1, 12, 4, cachedS[i]);
						gl.glEvalMesh2(type, 0, grid, 0, grid);
		  			}
		  		}
		    	gl.glPopMatrix();
		    	gl.glPopAttrib();
		  	}
		  	
			protected override void DrawObject()
			{
				bool solid = true;
				double scale = 0.5; 
		  
			  	if (solid)
			    	DoTeaPot (7, scale, gl.GL_FILL);
			  	else
				    DoTeaPot (10, scale, gl.GL_LINE);
			}
		}
}
